packages/core/src/proxy_configuration.ts

```diff
@@ -189,6 +267,60 @@ export class ProxyConfiguration {
         };
     }
 
+    /**
+     * Given a session identifier and a request / proxy tier, this function returns a new proxy URL based on the provided configuration options.
+     * @param _sessionId Session identifier
+     * @param options Options for the tiered proxy rotation
+     * @returns A string with a proxy URL.
+     */
+    protected _handleTieredUrl(_sessionId: string, options?: TieredProxyOptions): string {
+        if (!this.tieredProxyUrls) throw new Error('Tiered proxy URLs are not set');
+
+        if (!options || (!options?.request && options?.proxyTier === undefined)) {
+            const allProxyUrls = this.tieredProxyUrls.flat();
+            return allProxyUrls[this.nextCustomUrlIndex++ % allProxyUrls.length];
+        }
+
+        let tierPrediction = options.proxyTier!;
+
+        if (typeof tierPrediction !== 'number') {
+            tierPrediction = this.getProxyTier(options.request!)!;
+        }
+
+        const proxyTier = this.tieredProxyUrls![tierPrediction];
```

Comment:
I'm surprised you need a null assertion here

Reply:
So was I, it seems to be something about the `this` reference - you don't need it here if `this.tieredProxyUrls` was a local variable. Weird indeed 🤷🏽